js/customize-snapshots.js

```diff
@@ -731,16 +701,99 @@
 		 * @returns {string} URL.
 		 */
 		getSnapshotFrontendPreviewUrl: function getSnapshotFrontendPreviewUrl() {
-			var snapshot = this, a = document.createElement( 'a' );
+			var snapshot = this, a = document.createElement( 'a' ),
+				params = {};
+
+			if ( api.settings.changeset && api.settings.changeset.uuid ) {
+				params.customize_changeset_uuid = api.settings.changeset.uuid;
+			} else {
+				params.customize_changeset_uuid = snapshot.data.uuid;
+			}
+
 			a.href = snapshot.frontendPreviewUrl.get();
-			if ( a.search ) {
-				a.search += '&';
+			if ( snapshot.statusButton.disbleButton.get() ) {
+				return a.href;
+			}
+			if ( ! api.settings.theme.active ) {
+				params.theme = api.settings.theme.stylesheet;
 			}
-			a.search += snapshot.uuidParam + '=' + snapshot.data.uuid;
+			a.search = $.param( params );
+
 			return a.href;
 		},
 
 		/**
+		 * Frontend preview link setup.
+		 *
+		 * @returns {void}
+		 */
+		setUpPreviewLink: function setUpPreviewLink() {
+			var snapshot = this, setPreviewLinkHref;
+
+			snapshot.previewLink = $( $.trim( wp.template( 'snapshot-preview-link' )() ) );
+			snapshot.previewLink.attr( 'target', snapshot.data.uuid );
+			snapshot.previewLink.href = snapshot.frontendPreviewUrl.get();
+
+			setPreviewLinkHref = _.debounce( function() {
+				var queryVars;
+
+				snapshot.updateSnapshotPreviewLinkHref( false );
+
+				// Add the customize_theme param to the frontend URL if the theme is not active.
+				if ( ! api.state( 'activated' ).get() ) {
+					queryVars = snapshot.parseQueryString( snapshot.previewLink.prop( 'search' ).substr( 1 ) );
+					queryVars.customize_theme = api.settings.theme.stylesheet;
+					snapshot.previewLink.prop( 'search', $.param( queryVars ) );
+				}
+			} );
+
+			snapshot.previewLink.click( function( e ) {
+				e.preventDefault();
+				snapshot.updateSnapshotPreviewLinkHref( true );
+			} );
+
+			snapshot.frontendPreviewUrl.bind( setPreviewLinkHref );
+			api.state.bind( 'change', setPreviewLinkHref );
+			api.bind( 'saved', setPreviewLinkHref );
+			$( '#customize-footer-actions .devices button' ).first().before( snapshot.previewLink );
+		},
+
+		/**
+		 * Updates the snapshot preview link href and opens the link when defined.
+		 *
+		 * @param {boolean} openLink If to open the link.
+		 * @returns {void}
+		 */
+		updateSnapshotPreviewLinkHref: function updateSnapshotPreviewLinkHref( openLink ) {
+			var onceProcessingComplete,
+				snapshot = this;
+
+			onceProcessingComplete = function() {
+				var request, href;
+				if ( api.state( 'processing' ).get() > 0 ) {
+					return;
+				}
+
+				api.state( 'processing' ).unbind( onceProcessingComplete );
+				request = api.requestChangesetUpdate();
+
+				request.done( function() {
+					href = snapshot.getSnapshotFrontendPreviewUrl();
+					snapshot.previewLink.attr( 'href', href );
+					if ( openLink ) {
+						window.open( href );
```

Comment:
And I think because of `window.open` I get popup blocked, when I click on the button. 

![popup-blocked](https://user-images.githubusercontent.com/6297436/30774142-64894b92-a09b-11e7-8970-f639d906fa12.png)

Steps to produce: 
- Open fresh customizer without changeset uuid
- Make some changes ( I made in site title )
- Now click on the button.

Reply:
@westonruter

Reply:
This doesn't look like an ideal way to doing it. The clicking should be natural. If there are dirty changes may be let the user know through some notification like "Preparing preview.." and then let them click. Just an idea.

Reply:
@westonruter  Another idea would be that when you mouse enter the publish settings section, or preview link control container, we execute `api.requestChangesetUpdate()` and if there are dirty changes to be saved, make the link disabled, show a notification or loader, "Preparing preview..", and then enable the preview link when done. Thoughts?

Reply:
@sayedtaqui yes, I like that idea. If there are unsaved changes, then disable the area until you've saved them. In fact, we could actually make the appearance of the preview link conditional on you having saved changes to begin with. We wouldn't want someone try to share a link to preview an _autosaved_ state. So maybe the preview link control could appear disabled with a notification, “Save your changes in order to share the preview.” In that way you wouldn't need to have a “Preparing preview” message or any attempt to save the changeset just-in-time when clicking the link.

Reply:
> preview link control could appear disabled with a notification, “Save your changes in order to share the preview.”
I think this should be the best way to do it. Because if the changes are not saved as draft/future, the share link might not be useful for the person they have shared it with, because the changeset may not even exist if they have clicked on the x icon on reload or it was garbage collected ( auto-draft )